calyx-opt/src/passes/compile_invoke.rs

```diff
@@ -131,38 +132,59 @@ impl CompileInvoke {
     ///
     /// Since this pass eliminates all ref cells in post order, we expect that
     /// invoked component already had all of its ref cells removed.
+
     fn ref_cells_to_ports<T>(
         &mut self,
         inv_cell: RRC<ir::Cell>,
         ref_cells: impl Iterator<Item = (ir::Id, ir::RRC<ir::Cell>)>,
     ) -> Vec<ir::Assignment<T>> {
         let inv_comp = inv_cell.borrow().type_name().unwrap();
         let mut assigns = Vec::new();
-        for (ref_cell_name, cell) in ref_cells {
+        for (ref_cell_name, in_cell) in ref_cells {
             log::debug!(
                 "Removing ref cell `{}` with {} ports",
                 ref_cell_name,
-                cell.borrow().ports.len()
+                in_cell.borrow().ports.len()
             );
 
             // Mapping from canonical names of the ports of the ref cell to the
-            // new port defined on the signature of the component
+            // new port defined on the signature of the component. This has name of ref cell, not arg cell
             let Some(comp_ports) = self.port_names.get(&inv_comp) else {
                 unreachable!("component `{}` invoked but not already visited by the pass", inv_comp)
             };
 
-            // The type of the cell is the same as the ref cell so we can
-            // iterate over its ports and generate bindings for the ref cell.
-            for pr in &cell.borrow().ports {
+            // We expect each canonical port in `comp_ports` to exactly match with a port in
+            //`in_cell` based on well-formedness subtype checks.
+            for canon in comp_ports.keys() {
+                //only interested in ports attached to the ref cell
+                if canon.cell != ref_cell_name {
+                    continue;
+                }
+                let in_cell_borrowed = in_cell.borrow();
+                // The given port of actual, concrete cell passed in
+                let pr = in_cell_borrowed
+                    .ports
+                    .iter()
+                    .find(|&in_cell_port| {
+                        canon.cell == ref_cell_name
+                            && in_cell_port.borrow().name == canon.port
+                    })
+                    .unwrap_or_else(|| {
+                        unreachable!(
+                            "port `{}` not found in the cell `{}`",
+                            canon,
+                            in_cell.borrow().name()
+                        )
+                    });
```

Comment:
Maybe it would be helpful to outline this to a function of its own… named something like `find_cell_port`, to make it clear that we are looking for the corresponding port on the passed-in cell.

Reply:
This is totally a non-urgent self-nit, but I find this stuff interesting because in general I struggle with idiomatic rust and want to get better at it.

A possible implementation could look like this (note in the signature we expect an `RRC`):

```rust
 fn get_concrete_port(concrete_cell : RRC<ir::Cell>, canonical_port : &ir::Id) -> RRC<ir::Port>{
        let concrete_cell = concrete_cell.borrow();
        concrete_cell.ports.iter().find(|&concrete_cell_port| {
            concrete_cell_port.borrow().name == canonical_port
        }).unwrap_or_else(|| {
            unreachable!(
                "port `{}` not found in the cell `{}`",
                canonical_port,
                concrete_cell.name()
            )
        }).clone()
    }
}
```

With a function call like this
```rust
let concrete_port = Self::get_concrete_port(concrete_cell.clone(), &canon.port);
```

Another option is to have a signature like this where instead of an `RRC` a "raw" reference is expected
```rust
fn get_concrete_port(concrete_cell : &ir::Cell, canonical_port : &ir::Id) -> RRC<ir::Port>{
```

with a function call like this

```rust
let concrete_port = Self::get_concrete_port(&concrete_cell.borrow(), &canon.port);
```

Curious if one is "better" than the other? My gut is that if we're already using `RefCell`s we might as well keep them around, and so the first option is more in line with that.